File: core/src/main/java/nl/t64/game/rpg/ResourceManager.java
package nl.t64.game.rpg;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.assets.AssetManager;
import com.badlogic.gdx.assets.loaders.TextureLoader;
import com.badlogic.gdx.files.FileHandle;
import com.badlogic.gdx.graphics.Texture;
import com.badlogic.gdx.graphics.g2d.BitmapFont;
import com.badlogic.gdx.graphics.g2d.TextureAtlas;
import com.badlogic.gdx.graphics.g2d.TextureRegion;
import com.badlogic.gdx.graphics.g2d.freetype.FreeTypeFontGenerator;
import com.badlogic.gdx.graphics.g2d.freetype.FreeTypeFontGeneratorLoader;
import com.badlogic.gdx.graphics.g2d.freetype.FreetypeFontLoader;
import com.badlogic.gdx.maps.tiled.TiledMap;
import com.badlogic.gdx.maps.tiled.TmxMapLoader;
import com.badlogic.gdx.utils.Json;
import com.badlogic.gdx.utils.ObjectMap;
import nl.t64.game.rpg.conversation.ConversationGraph;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;


public class ResourceManager {

    private static final String SPRITE_CONFIGS = "configs/sprites";
    private static final String CONVERSATION_CONFIGS = "configs/conversations/";
    private static final String SHOP_CONFIGS = "configs/shops/";
    private static final String CONFIG_SUFFIX = ".json";
    private static final String ATLAS_FILES = "sprites";
    private static final String ATLAS_FILES2 = "sprites/inventory";
    private static final String ATLAS_SUFFIX = ".atlas";

    private final AssetManager assetManager;
    private final ObjectMap<String, SpriteConfig> spriteConfigs;
    private final List<TextureAtlas> atlasList;
    private final Json json;

    public ResourceManager() {
        this.assetManager = new AssetManager();
        this.spriteConfigs = new ObjectMap<>();
        this.atlasList = new ArrayList<>();
        this.json = new Json();
    }

    public void unloadAsset(String assetFilenamePath) {
        if (assetManager.isLoaded(assetFilenamePath)) {
            assetManager.unload(assetFilenamePath);
        }
    }

    public TiledMap getMapAsset(String mapFilenamePath) {
        if (!assetManager.isLoaded(mapFilenamePath)) {
            loadMapAsset(mapFilenamePath);
        }
        return assetManager.get(mapFilenamePath, TiledMap.class);
    }

    private void loadMapAsset(String mapFilenamePath) {
        assetManager.setLoader(TiledMap.class, new TmxMapLoader(assetManager.getFileHandleResolver()));
        assetManager.load(mapFilenamePath, TiledMap.class);
        assetManager.finishLoadingAsset(mapFilenamePath);
    }

    public BitmapFont getTrueTypeAsset(String trueTypeFilenamePath, int fontSize) {
        if (!assetManager.isLoaded(trueTypeFilenamePath)) {
            loadTrueTypeAsset(trueTypeFilenamePath, fontSize);
        }
        return assetManager.get(trueTypeFilenamePath, BitmapFont.class);
    }

    private void loadTrueTypeAsset(String trueTypeFilenamePath, int fontSize) {
        assetManager.setLoader(FreeTypeFontGenerator.class, new FreeTypeFontGeneratorLoader(assetManager.getFileHandleResolver()));
        assetManager.setLoader(BitmapFont.class, ".ttf", new FreetypeFontLoader(assetManager.getFileHandleResolver()));
        var parameter = new FreetypeFontLoader.FreeTypeFontLoaderParameter();
        parameter.fontFileName = trueTypeFilenamePath;
        parameter.fontParameters.size = fontSize;
        assetManager.load(trueTypeFilenamePath, BitmapFont.class, parameter);
        assetManager.finishLoadingAsset(trueTypeFilenamePath);
    }

    public Texture getTextureAsset(String textureFilenamePath) {
        if (!assetManager.isLoaded(textureFilenamePath)) {
            loadTextureAsset(textureFilenamePath);
        }
        return assetManager.get(textureFilenamePath, Texture.class);
    }

    private void loadTextureAsset(String textureFilenamePath) {
        assetManager.setLoader(Texture.class, new TextureLoader(assetManager.getFileHandleResolver()));
        assetManager.load(textureFilenamePath, Texture.class);
        assetManager.finishLoadingAsset(textureFilenamePath);
    }

    public SpriteConfig getSpriteConfig(String spriteId) {
        if (spriteConfigs.size == 0) {
            loadSpriteConfigs();
        }
        return spriteConfigs.get(spriteId);
    }

    @SuppressWarnings("unchecked")
    private void loadSpriteConfigs() {
        FileHandle[] configFiles = Gdx.files.local(SPRITE_CONFIGS).list(CONFIG_SUFFIX);
        for (FileHandle file : configFiles) {
            ObjectMap<String, SpriteConfig> configs = json.fromJson(ObjectMap.class, SpriteConfig.class, file);
            spriteConfigs.putAll(configs);
        }
    }

    public TextureRegion getAtlasTexture(String atlasId) {
        if (atlasList.isEmpty()) {
            loadAtlasTexture(ATLAS_FILES);
            loadAtlasTexture(ATLAS_FILES2);
        }
        TextureRegion textureRegion = null;
        for (TextureAtlas atlas : atlasList) {
            textureRegion = atlas.findRegion(atlasId);
            if (textureRegion != null) {
                break;
            }
        }
        return textureRegion;
    }

    private void loadAtlasTexture(String path) {
        FileHandle[] atlasFiles = Gdx.files.local(path).list(ATLAS_SUFFIX);
        Arrays.stream(atlasFiles)
              .map(TextureAtlas::new)
              .forEach(atlasList::add);
    }

    public ConversationGraph getConversation(String conversationId) {
        String fullFilenamePath = CONVERSATION_CONFIGS + conversationId + CONFIG_SUFFIX;
        return json.fromJson(ConversationGraph.class, Gdx.files.local(fullFilenamePath));
    }

    @SuppressWarnings("unchecked")
    public List<String> getShopInventory(String shopId) {
        String fullFilenamePath = SHOP_CONFIGS + shopId + CONFIG_SUFFIX;
        return json.fromJson(List.class, String.class, Gdx.files.local(fullFilenamePath));
    }

}

File: core/src/nl/t64/game/rpg/ResourceManager.kt
package nl.t64.game.rpg

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.assets.AssetLoaderParameters
import com.badlogic.gdx.assets.AssetManager
import com.badlogic.gdx.assets.loaders.MusicLoader
import com.badlogic.gdx.assets.loaders.SoundLoader
import com.badlogic.gdx.assets.loaders.TextureLoader
import com.badlogic.gdx.audio.Music
import com.badlogic.gdx.audio.Sound
import com.badlogic.gdx.graphics.Texture
import com.badlogic.gdx.graphics.g2d.BitmapFont
import com.badlogic.gdx.graphics.g2d.TextureAtlas
import com.badlogic.gdx.graphics.g2d.TextureRegion
import com.badlogic.gdx.graphics.g2d.freetype.FreeTypeFontGeneratorLoader
import com.badlogic.gdx.graphics.g2d.freetype.FreetypeFontLoader
import com.badlogic.gdx.maps.tiled.TiledMap
import com.badlogic.gdx.maps.tiled.TmxMapLoader
import com.badlogic.gdx.utils.Json
import ktx.assets.getAsset
import ktx.assets.loadOnDemand
import ktx.assets.setLoader
import ktx.assets.unloadSafely
import ktx.collections.GdxMap
import ktx.json.fromJson


private const val FILE_LIST = "_files.txt"
private const val SPRITE_CONFIGS = "configs/sprites/"
private const val FILE_LIST_SPRITE_CONFIGS = SPRITE_CONFIGS + FILE_LIST
private const val SHOP_CONFIGS = "configs/shops/"
private const val CONFIG_SUFFIX = ".json"
private const val ATLAS_FILES = "sprites/"
private const val FILE_LIST_ATLAS_FILES = ATLAS_FILES + FILE_LIST
private const val ATLAS_FILES2 = "sprites/inventory/"
private const val FILE_LIST_ATLAS_FILES2 = ATLAS_FILES2 + FILE_LIST
private const val ATLAS_FILES3 = "sprites/spells/"
private const val FILE_LIST_ATLAS_FILES3 = ATLAS_FILES3 + FILE_LIST

class ResourceManager {

    private val assetManager = AssetManager().apply {
        setLoader(TmxMapLoader(fileHandleResolver))
        setLoader(FreeTypeFontGeneratorLoader(fileHandleResolver))
        setLoader(FreetypeFontLoader(fileHandleResolver), ".ttf")
        setLoader(TextureLoader(fileHandleResolver))
        setLoader(SoundLoader(fileHandleResolver))
        setLoader(MusicLoader(fileHandleResolver))
    }
    private val spriteConfigs = GdxMap<String, SpriteConfig>()
    private val atlasList = ArrayList<TextureAtlas>()
    private val json = Json()

    fun unloadAsset(assetFilenamePath: String) {
        assetManager.unloadSafely(assetFilenamePath)
    }

    fun getMapAsset(mapFilenamePath: String): TiledMap {
        return getAsset(mapFilenamePath)
    }

    fun getTrueTypeAsset(trueTypeFilenamePath: String, fontSize: Int): BitmapFont {
        val parameters = FreetypeFontLoader.FreeTypeFontLoaderParameter()
        parameters.fontFileName = trueTypeFilenamePath
        parameters.fontParameters.size = fontSize
        return getAsset(trueTypeFilenamePath, parameters)
    }

    fun getTextureAsset(textureFilenamePath: String): Texture {
        return getAsset(textureFilenamePath)
    }

    fun getSoundAsset(soundFilenamePath: String): Sound {
        return getAsset(soundFilenamePath)
    }

    fun getMusicAsset(musicFilenamePath: String): Music {
        return getAsset(musicFilenamePath)
    }

    private inline fun <reified T : Any> getAsset(path: String, parameters: AssetLoaderParameters<T>? = null): T {
        assetManager.loadOnDemand(path, parameters).finishLoading()
        return assetManager.getAsset(path)
    }

    fun getSpriteConfig(spriteId: String): SpriteConfig {
        if (spriteConfigs.isEmpty) {
            loadSpriteConfigs()
        }
        return spriteConfigs[spriteId]
    }

    private fun loadSpriteConfigs() {
        Gdx.files.internal(FILE_LIST_SPRITE_CONFIGS).readString()
            .split(System.lineSeparator())
            .map { Gdx.files.internal(SPRITE_CONFIGS + it) }
            .map { json.fromJson(GdxMap::class.java, SpriteConfig::class.java, it) }
            .forEach { spriteConfigs.putAll(it as GdxMap<String, SpriteConfig>) }
    }

    fun getAtlasTexture(atlasId: String): TextureRegion {
        if (atlasList.isEmpty()) {
            loadAtlasTexture(FILE_LIST_ATLAS_FILES, ATLAS_FILES)
            loadAtlasTexture(FILE_LIST_ATLAS_FILES2, ATLAS_FILES2)
            loadAtlasTexture(FILE_LIST_ATLAS_FILES3, ATLAS_FILES3)
        }
        return atlasList.mapNotNull { it.findRegion(atlasId) }.first()
    }

    private fun loadAtlasTexture(fileName: String, directory: String) {
        Gdx.files.internal(fileName).readString()
            .split(System.lineSeparator())
            .filter { it.isNotBlank() }
            .map { Gdx.files.internal(directory + it) }
            .map { TextureAtlas(it) }
            .forEach { atlasList.add(it) }
    }

    fun getShopInventory(shopId: String): List<String> {
        val fullFilenamePath = SHOP_CONFIGS + shopId + CONFIG_SUFFIX
        return json.fromJson(Gdx.files.internal(fullFilenamePath))
    }

}
